PopPUNK/lineage_clustering.py

```diff
@@ -0,0 +1,680 @@
+    distances = X[:,distance_index]
+    
+    # sort distances
+    distance_ranks = np.argsort(distances)
+    distances = distances[distance_ranks]
+    
+    # determine whether ref-ref or ref-query analysis
+    isolate_list = rlist
+    
+    # determine whether novel analysis or modifying existing analysis
+    use_existing = False
+    neighbours = {}
+    lineage_seed = {}
+    lineage_clustering = {}
+    previous_lineage_clustering = {}
+    
+    null_cluster_value = len(isolate_list) + 1
+    for R in rank_list:
+        lineage_clustering[R] = {i:null_cluster_value for i in isolate_list}
+        lineage_seed[R] = {}
+        neighbours[R] = {}
+        previous_lineage_clustering[R] = {}
+    
+    # shared memory data structures
+    shm_distances = shared_memory.SharedMemory(create = True, size = distances.nbytes, name = 'shm_distances')
+    distances_shared = np.ndarray(distances.shape, dtype = distances.dtype, buffer = shm_distances.buf)
+    distances_shared[:] = distances[:]
+
+    shm_distance_ranks = shared_memory.SharedMemory(create = True, size = distance_ranks.nbytes, name = 'shm_distance_ranks')
+    distance_ranks_shared = np.ndarray(distance_ranks.shape, dtype = distance_ranks.dtype, buffer = shm_distance_ranks.buf)
+    distance_ranks_shared[:] = distance_ranks[:]
+
+    isolate_list_shared = shared_memory.ShareableList(isolate_list, name = 'shm_isolate_list')
+    
+    # run clustering for an individual R
+    if num_processes == 1:
+        for R in rank_list:
+            lineage_clustering[R], lineage_seed[R], neighbours[R], previous_lineage_clustering[R] = run_clustering_for_R(R,
+                                                                                        null_cluster_value = null_cluster_value,
+                                                                                        qlist = qlist,
+                                                                                        existing_scheme = existing_scheme,
+                                                                                        distances_length = distances.shape,
+                                                                                        distances_type = distances.dtype)
+
+
+    else:
+
+        with Pool(processes = num_processes) as pool:
+            results = pool.map(partial(run_clustering_for_R,
+                                null_cluster_value = null_cluster_value,
+                                qlist = qlist,
+                                existing_scheme = existing_scheme,
+                                distances_length = distances.shape,
+                                distances_type = distances.dtype),
+                                rank_list)
+        
+        for n,result in enumerate(results):
+            R = rank_list[n]
+            lineage_clustering[R], lineage_seed[R], neighbours[R], previous_lineage_clustering[R] = result
+        
+    # manage memory
```

Comment:
Fun that python now has us needing to `free`. Will the garbage collector do this for us at the end of the function?

Reply:
Solved with manager in da5b6df826fe1cafe622094f82cb8d3677354661.